Object Avoidance Database

magicrub opened the pull request:
Store object avoidance data from the Proximity library into a local database and feed them into the other Object Avoidance libraries as mini fences.

Two videos:
https://photos.app.goo.gl/QL58zGF3UHHE9pEB7

https://photos.app.goo.gl/wQNtRtdCRTagQysZ8

The circles drawn on MAVProxy in the videos are a slight mod on adsb msgs using an out-of-range emitter_type of 255. PR https://github.com/ArduPilot/MAVProxy/pull/643 adds support for that.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        // index out of range
+        return false;
+    }
+    if (_database.items[index].loc.lat == item.loc.lat &&
```
The odds of this seem so low why not go directly to the range calculation? If it's an exact match it's still going to pass the distance check, and if it wasn't an exact match this doesn't get you out early. Basically this only saves time when it's a perfect match, which is probably the least likely case, and then doing this while cycling through the full database is going to add up to be expensive.

rmackay9 replied:
we've removed the check so marking as resolved

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+
+    if (index != _database.count) {
+        // copy last object in array over expired object
+        _database.items[index] = _database.items[_database.count];
```
Presumably if last_hit was the last index you should also be updating here on this copy.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        // we failed to push, probably because it's full. If we're not at max size let's grow the queue size.
+        // we can't immediately deinit the queue because the background thread may be in the middle of emptying it
+        // so the defer sets a flag to let the background thread do the actual deinit
+        deinit_queue(true);
```
If you deinit the queue from anywhere other then the avoidance thread/update enviroment then you can get a nullptr dereference within Avoidance update.

rmackay9 replied:
deinit_queue is gone now so marking as resolved.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+
+        const uint32_t icao = index;
+
+        mavlink_msg_adsb_vehicle_send(chan,
```
I'm hoping this is just a temp version of this, and a proper message will be found?

rmackay9 replied:
yes, we will eventually create a new message to send the obstacle data to the GCS.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.h:
```diff
@@ -0,0 +1,111 @@
+    AP_OADatabase &operator=(const AP_OADatabase&) = delete;
+
+    struct OA_DbItem {
+        Location loc;
```
For what you need Location is a very expensive object to be storing in the queue. All you need is 8 bytes for lat/lng, and instead you are getting all the flags + altitude coming along with it. For what you are doing it's probably better to store this as 2 int32_t's and make a distance helper that can do math directly on that. (Or alternatively Vector2l). This would cut the cost of these structs from 20 bytes per, to 12 bytes per. This is a considerable size difference, which makes this a lot easier to fit on the embedded boards, and on anything that has a cache (the M7's and up) a lot faster.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.h:
```diff
@@ -0,0 +1,111 @@
+        const float     filter_min_m            = 0.011f; // worst case resolution of int32 lat/lng value at equator is 1.1cm;
+        const float     filter_grow_rate        = 1.03f;
+        const float     filter_shrink_rate      = 0.99f;
+        uint32_t        count                   = 0;        // items count
```
Unneeded initilizer

rmackay9 replied:
all non-zero initialisers have been removed.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+    _database.size_param_last = _database.size_param;
+
+    _database.size = 0;
+    _database.count = 0;
```
Should reset the last hit index.

rmackay9 replied:
last hit index removed so marking as resolved

WickedShell commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -198,6 +206,7 @@ void AP_OAPathPlanner::avoidance_thread()
         bool res = false;
         switch (_type) {
         case OA_PATHPLAN_DISABLED:
+        default:
```
Not having the default meant that the compiler could warn you if you added a case that wasn't covered here. In general I actually like not including default when we get into large switches like this.

magicrub replied:
fixed

WickedShell commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -206,14 +215,15 @@ void AP_OAPathPlanner::avoidance_thread()
             res = _oadijkstra->update(avoidance_request2.current_loc, avoidance_request2.destination, origin_new, destination_new);
             break;
-        }
+        } // switch
```
Kinda a messy/low value comment...

magicrub replied:
removed

WickedShell commented on libraries/AP_Proximity/AP_Proximity_Backend.cpp:
```diff
@@ -351,40 +330,32 @@ bool AP_Proximity_Backend::get_next_ignore_start_or_end(uint8_t start_or_end, in
 
-void AP_Proximity_Backend::update_locations()
+// returns true if database is ready to be pushed to and all cached data is ready
+bool AP_Proximity_Backend::database_prepare_for_push()
```
I'm struggling to see the value of having this spun out like this. Why not just do it inside the push as an early return? Seems more complicated.

magicrub replied:
reworked, thanks for the suggetsion

WickedShell commented on libraries/AP_Proximity/AP_Proximity_Backend.cpp:
```diff
@@ -351,40 +330,32 @@ bool AP_Proximity_Backend::get_next_ignore_start_or_end(uint8_t start_or_end, in
-    }
+    // current_loc and current_vehicle_bearing have
+    // been cached from database_prepare_for_push()
+    Location temp_loc = current_loc;
```
If prepare got brought inside this function then the cached current_loc can be removed which saves ram/copying and clearer.

magicrub replied:
Most of the Proxy drivers do this, but there's a MAVLink msg that would run this 72 times so we don't need to recacl the heading 72 times.

WickedShell commented on libraries/AP_Proximity/AP_Proximity_Backend.h:
```diff
@@ -112,8 +107,7 @@ class AP_Proximity_Backend
-    AP_Proximity::Proximity_Location _locations[PROXIMITY_SECTORS_MAX];   // buffer of locations
-    HAL_Semaphore_Recursive _rsem;                          // semaphore for access to _locations and _location_count
+    // database cached values to speed things up
+    float current_vehicle_bearing;
```
Same with this, if inlined can disappear.

magicrub replied:
fixed

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+
+void AP_OADatabase::optimize_db_filter()
+{
+    process_stats("\nOA queue", _queue.stats, _queue.items->available());
```
yeah, lots of printing going on. I need to surround it all with an #if DEBUG.

magicrub replied:
fixed

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+
+#define AP_OADATABASE_TIMEOUT_SECONDS_DEFAULT   10
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
```
fixed

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        return;
+    }
+
+    delete []_database.items;
```
fixed

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        } else {
+            // failed to allocate, try a little smaller next time
+            ::printf("DB failed to allocate database of size %d\n", _database.size);
+            _database.size_param.set_and_notify(_database.size_param * 0.9);
```
fixed

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        // we're almost full, lets increase the filter size by requiring more
+        // spacing between points so less things get put into the database
+        if (_database.filter_m < _database.filter_max_m) {
+            _database.filter_m = constrain_float(_database.filter_m*_database.filter_grow_rate,_database.filter_min_m, _database.filter_max_m);
```
fixed

rmackay9 commented:
A few points from a quick review:

- it looks like we are trying to support re-allocating the size of the database in real-time.  I really don't think this is necessary nor a good idea.  Let's just add RebootRequired to the DB_SIZE parameter's description and save ourselves the code complexity and danger of real-time resizing
- I don't think the outputting to ADSB can go into master in this form at least.  The GCS object is really responsible for throttling messages and giving the user control of the rates but this is completely bypassed as it is in this PR.  I'm afraid that for now at least we should just remove this feature for the sake of getting this PR merged.  If we really want to get some reporting in then we should add a "send_adsb_vehicle()" function similar to AP_ADSB's. 
- I think we should remove most of the debug.  It's useful for us as devs while we are working on this but not sure it's good to leave in as we merge to master.  I vote for removing it all and then add a "healthy" accessor.  We will disable object avoidance if the database becomes "unhealthy".
- AP_OADatabase::get_database() and AP_OAPathPlanner::get_database() are never called so they can be removed
- Some style and formatting issues:
  - in AP_OADatabase.h, let's complete the comments and align them properly.  for example some _database fields hae comments, others don't.  it would be good to add comments above "process_stats" and the structure above it.
  - in _database, let's not mix the parameters in with the local variables.  let's either move them out of the _database structure or put them at the top of the structure's definition.  I would prefer we move all parameters out and group them together.
  - can we remove or rename "service_output_throttle" and maybe add some comments in the .h file?
  - can we reorder the implementations in the .cpp to match the order in the .h?
  - "averge" should be "average"

magicrub commented:
@rmackay9 all suggestions have been applied and I've removed the WIP label. Also squashed and rebased. This is ready to be merged. My testing consists of "well, it still looks the same on Morse" but I have zero debug feedback to know what its thinking so I have to just assume it still works the same way.

The useful debug branch is https://github.com/magicrub/ardupilot/tree/avoidance-database-ExtraStuff which is the same PR commit before I ripped out the deinit_database and adsb send stuff.

magicrub commented:
What race condition?

magicrub commented:
@WickedShell my apologies, I only saw half of your comments. I must have not refreshed

magicrub commented:
I ripped out more useful stuff, no more last_hit, no more queue deinit. We now fail fast and permanently.

rmackay9 commented:
@magicrub, 

A few more things I've noticed are below.  Sorry to keep piling on issues but I'm just looking to help make it as solid as possible so we can get it in.

- I think AP_OADatabase's init() function should be made public and then be called from AP_OAPathPlanner's init function instead of from it's own "update()" function.  This ensures we do the memory location during initialisation not while the vehicle is in the air.  It will also allow, "_unable_to_allocate_memory" to be removed because init() will never be called more than once.
- in database_remove_item() I don't think these two lines are necessary:
>    // invalidate the location so that a UI will redraw it
>    _database.items[index] = _database.invalid_item;
- AP_OAPathPlanner::database_healthy() and is_item_in_database() can be made const
- I wonder if perhaps AC_Proximity should be pushing locations directly into the AP_OADatabase instead of going through the AP_OAPathPlanner object.
- I think for now, before we have any real information on the performance differences of different flight controllers we should just define AP_OADATABASE_SIZE_DEFAULT to 100 (or whatever).  Also i think we want SITL to perform like a real vehicle rather than perform much better.
- "is_item_in_database" could be renamed to perhaps, "close_to_item_in_database".
- some minor formatting issues:
  - there's a single character space missing for "size_min"
  - comments are missing for "filter_grow_rate", "filter_shrink_rate", "optimize_last_ms"

rmackay9 commented:
one more thing, in AP_Proximity_Backend::update_boundary_for_sector() there's no need to make the pass-by-value arguments (a uint8_t and a bool) const.  There's nothing inherently wrong with the function internally modifying those arguments and I worry it will lead to this becoming some kind of standard.

magicrub commented:
>>I worry it will lead to this becoming some kind of standard
Too late, it's already a standard! And a good one too. It's called "const correctness". https://isocpp.org/wiki/faq/const-correctness

magicrub commented:
everything has been address again. I think the database is about a third the size of the fully featured one. The database is about as simple as we can get it. I'd like not dumb it down any more, it's horrible now. I can't test anything. It "just works" with zero feedback and zero insight. I am strongly against this type of merge. No one else is now able to develop usefully on it. I've spent about as much time ripping things out as I have putting them in.

magicrub commented:
>AC_Proximity should be pushing locations directly into the AP_OADatabase

I don't think so as long as AP_OAPathPlanner is the parent class of AP_OADatabase. perhaps @peterbarker would be the man to talk to about this, but should a child class deserve it's own singleton? I say no.

WickedShell commented:
> I'd like not dumb it down any more, it's horrible now

I wasn't suggesting to dumb it down on the nullptr one, you just needed a slightly wider semaphore scope + nullptr checks.

magicrub commented:
Can you be more specific regarding semephore use?

rmackay9 commented:
I'm definitely not a fan of adding const in front of arguments that are passed by value because it just adds more words that make the function longer and thus slightly harder to read.  I suspect other devs will agree.  Let's discuss on the dev call.

peterbarker commented:
On Sun, 30 Jun 2019, Tom Pittenger wrote:

> I don't think so as long as AP_OAPathPlanner is the parent class of AP_OADatabase. perhaps @peterbarker would be the man to talk to about this, but should
> a child class deserve it's own singleton? I say no.

AP::ahrs_navekf() vs AP::ahrs()

peterbarker commented:
On Sun, 30 Jun 2019, Randy Mackay wrote:

> I'm definitely not a fan of adding const in front of arguments that are passed by value because it just adds more words that make the function longer and
> thus slightly harder to read. I suspect other devs will agree. Let's discuss on the dev call.

I don't find it harder to read.

We const other local variables where we can - which is what these are.

Just don't const them in the header - that's not necessary as the caller 
knows the function isn't going to modify their copy.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,291 @@
+        for (uint32_t i=0; i<_database.count; i++) {
+            if (is_item_in_database(i, item)) {
+                database_item_refresh(i, item.timestamp);
+                return;
```
I don't think `return` is correct here. This makes you stop processing the outer loop, which means if there is a match on the first item in the queue, nothing else is pulled out, which introduces a large delay of ~100ms before we pull the next item out of the queue. If you are near a wall and using the MAVLink message this could easily lead to not updating your queue correctly.

magicrub replied:
good catch, but it's still getting served with the 1ms delay that you didn't notice and thought was a busy loop.

rmackay9 commented:
@peterbarker, I think the declaration and the implementation will need to match.  I haven't tested but I'd be surprised if c++ thinks they match if the const is only added to the implementation.

magicrub commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -172,15 +194,19 @@ bool AP_OAPathPlanner::mission_avoidance(const Location &current_loc,
-        // run at 10hz or less
-        hal.scheduler->delay(100);
+        const uint32_t now = AP_HAL::millis();
+        if (now - avoidance_latest_ms < 100) {
```
because _oadatabase.update() needs to run very fast and it makes things much simpler if we put it into the same thread. There's a 1ms sleep there.

peterbarker commented:
On Sun, 30 Jun 2019, Randy Mackay wrote:

> @peterbarker, I think the declaration and the implementation will need to match. I haven't tested but I'd be surprised if c++ thinks they match if the
> const is only added to the implementation.

Feel free to test (or just look for existing examples).

I was pleasantly surprised to find they don't need to match - at least 
in the case we care about :-)

https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_Stats/AP_Stats.h#L32
https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_Stats/AP_Stats.cpp#L122

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,291 @@
+    // @Param: QUEUE_SIZE
+    // @DisplayName: OADatabase queue maximum number of points
+    // @Description: OADatabase queue maximum number of points.
+    // @Range: 1 1000
```
sure, why not. I'll change it to 80. I have zero feedback to whats happening, now that all the debug is gone, so being 5 or 50 or 100 or 500 will result in what seems appears to be same behavior. <shrug> It either "just works" or it does something else.

magicrub replied:
fixed

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,291 @@
+
+void AP_OADatabase::update()
+{
+    // exit immediately if we're unable to allocate memory or have just allocated/deallocated
```
I removed this 10 hours before this review so not sure why it's showing here now.

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,291 @@
+
+void AP_OADatabase::init_queue()
+{
+    if (_queue.items != nullptr || _queue_size_param < _queue.size_min) {
```
I removed this 10 hours before this review so not sure why it's showing here now.

magicrub commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -172,15 +194,19 @@ bool AP_OAPathPlanner::mission_avoidance(const Location &current_loc,
 {
     while (true) {
+        _oadatabase.update();
+        hal.scheduler->delay(1);
```
I've added some logic to help

WickedShell replied:
Sorry thought the later if was a while for some reason.

tridge commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,255 @@
+        return;
+    }
+
+    _database.items = (OA_DbItem *)calloc(_database.size, sizeof(OA_DbItem));
```
it looks like we will use this array even if alloc fails?

magicrub replied:
fixed

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.h:
```diff
@@ -0,0 +1,86 @@
+#include <AP_Common/Location.h>
+#include <AP_Param/AP_Param.h>
+#include <AP_HAL/utility/RingBuffer.h>
+#include <GCS_MAVLink/GCS_MAVLink.h>
```
Can we remove this header dependency?

magicrub replied:
done

rmackay9 commented:
Some comments from the dev call:

- let's measure the amount of RAM and CPU consumed on Rover
- make sure it's not included for 1MB boards

magicrub commented:
TODO:
- free both buffers if either fail to allocate
- HAL_MINIMIZE

tridge commented:
this needs a HAL_MINIMIZE_FEATURES check

peterbarker commented:
On Mon, 1 Jul 2019, Randy Mackay wrote:

>  *  let's measure the amount of RAM and CPU consumed on Rover

It was also suggested that performance be checked with higher loop rates 
and computation requirements.  The former could be done by fiddling 
SCHED_RATE but the latter requires an armed Copter doing full fusion.

magicrub commented:
> this needs a HAL_MINIMIZE_FEATURES check

done

magicrub commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -125,20 +154,13 @@ bool AP_OAPathPlanner::mission_avoidance(const Location &current_loc,
-        }
-        _thread_created = true;
+    // make sure threads are running
+    if (!start_thread()) {
```
fixed

magicrub commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,291 @@
+#endif
+
+#ifndef AP_OADATABASE_QUEUE_SIZE_DEFAULT
+    #define AP_OADATABASE_QUEUE_SIZE_DEFAULT 50
```
changed to 80

magicrub commented:
wow, I have some impressive results!

Firstly, this thing has lousy accuracy if you have lousy gps accuracy. The gps error transfers directly into object location error. That was a known going into this, but I just wanted to make sure everyone else knew.

tested on a CubeBlack on an AION Robotics R1. The avg was from a micros time stamp before and after running the db.update(). The vehicle was not in AUTO so it was not running the path algorithms, this is just the time it takes to do all the db maintenance.

| shedRate | queue Size | db Size | free RAM | Avg time(us) |
| --- | --- | --- | --- | --- |
| 50 | 200 | 100 | 90248 | 6 |
| 50 | 80 | 1000 | 68864 | 600  |
| 400 | 200 | 100 | 90248 | 12 |
| 400 | 80 | 1000 | 68864 | 700  |


So, it seems we could easily do a database of size 1000 if we wanted to on any platform! Wow, crazy! I can do more testing before the next dev call, but I think cpu wise the database is looking quite good as-is. Let's keep the default size to 100 for now.

magicrub commented:
rebased

rmackay9 commented:
@magicrub, nice results and great that it could be tested on real hardware.  I agree that sticking with 100 objects is good for now.  I think with a large number of objects it will be the BendyRuler calcs that will see a large jump in CPU usage.

WickedShell commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,277 @@
+
+    if (_queue.items == nullptr || _database.items == nullptr) {
+        gcs().send_text(MAV_SEVERITY_INFO, "DB init failed . Sizes queue:%d, db:%d", _queue.size, _database.size);
+        free(_queue.items);
```
Mixed new/free again here. This should be delete, or the new should be a variation of malloc.

magicrub replied:
I'd like a second opinion

magicrub replied:
per the dev call, all calloc()'s were removed and everything now uses new and delete

WickedShell commented on libraries/AC_Avoidance/AP_OAPathPlanner.h:
```diff
@@ -88,7 +90,11 @@ class AP_OAPathPlanner {
     bool _thread_created;           // true once background thread has been created
     AP_OABendyRuler *_oabendyruler; // Bendy Ruler algorithm
     AP_OADijkstra *_oadijkstra;     // Dijkstra's algorithm
+    AP_OADatabase _oadatabase;      // Database of dynamic polygon fences
```
This isn't being minimized, which means the database is always present in the build, even though users aren't seeing the parameters. Effectively this creates the situation where you have saved very little flash, while still including the feature in a non configurable version.

magicrub replied:
an empty class is being created when minimized
https://github.com/ArduPilot/ardupilot/pull/11671/files#diff-d5888bea59318da2160f3dd6eb3471e1R93

This is done because you can't get a singleton to an unknown class and then even if it's false, try to call isHealthy() on it like what the proximity library does.

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,277 @@
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("QUEUE_SIZE", 3, AP_OADatabase, _queue_size_param, AP_OADATABASE_QUEUE_SIZE_DEFAULT),
+AP_GROUPEND
```
slight formatting issue here with AP_GROUPEND

magicrub replied:
fixed

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,277 @@
+        free(_database.items);
+        return;
+    }
+    gcs().send_text(MAV_SEVERITY_INFO, "DB initialised. Sizes queue:%d, db:%d", _queue.size, _database.size);
```
I don't think we need this info being sent out when everything is working.  Imagine if every feature sent out a message as it initialised.

magicrub replied:
removed

rmackay9 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -112,6 +121,29 @@ bool AP_OAPathPlanner::pre_arm_check(char *failure_msg, uint8_t failure_msg_len)
     return true;
 }
 
+bool AP_OAPathPlanner::start_thread()
```
the start_thread() function seems to have some duplicate code that's already in mission_avoidance.  I think we should remove one of them (probably the one in mission_avoidance).

magicrub replied:
ahh yeah, I missed that when rebasing when you added that handy OA_ERROR result thing. It's also unreachable now too. Anyways, removed.

rmackay9 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -180,15 +212,31 @@ AP_OAPathPlanner::OA_RetState AP_OAPathPlanner::mission_avoidance(const Location
 void AP_OAPathPlanner::avoidance_thread()
 {
     while (true) {
+#if !HAL_MINIMIZE_FEATURES
```
It's not a blocker but I don't think the HAL_MINIMIZE_FEATURES should apply to only the object database.  I think we should disable all of the object avoidance.  The way it is here, bendyruler will avoid the fence but not dynamic obstacles.  I don't think having only half the feature isn't a good idea.

magicrub replied:
very much agreed. It was silly to place the requirement on just this part of the memory hungry avoidance. We can include the whole avoidance thing in HAL_MINIMIZE_FEATURES in another PR

rmackay9 replied:
great.  I've created an issue here so we don't forget about it. https://github.com/ArduPilot/ardupilot/issues/11770

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.h:
```diff
@@ -0,0 +1,106 @@
+    bool healthy() const { return _initialised; }
+
+    // fetch an item in database. Undefined result when i >= _database.count.
+    const OA_DbItem get_item(uint32_t i) const { return _database.items[i]; }
```
This should be const OA_DbItem& get_item(... to avoid copying.

magicrub replied:
yes, good idea. Missed that. Fixed!

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.h:
```diff
@@ -0,0 +1,106 @@
+    struct {
+        OA_DbItem       *items;                             // THE DATABASE
+        uint32_t        optimize_last_ms;
+        float           filter_m                = 1.0f;     // object avoidance database optimization level radius. Min distance between each fence point. Larger means lower resolution
```
It's not a blocker but I'm not sure that "filter_m" is a great name.  I think "radius" should be included in the name somehow.  Normally we don't add the units at the end if it's SI already.  Anyway.. not a huge deal but perhaps at some point we can come up with a better name.

magicrub replied:
active_radius or current_radius is probably better. next PR for that stuff

rmackay9 commented on libraries/AC_Avoidance/AP_OAPathPlanner.cpp:
```diff
@@ -180,15 +212,31 @@ AP_OAPathPlanner::OA_RetState AP_OAPathPlanner::mission_avoidance(const Location
+        }
 
-        // run at 10hz or less
+        const uint32_t now = AP_HAL::millis();
```
Again, not a blocker but let's remove the two lines below:

> hal.scheduler->delay(100);
> const uint32_t now = AP_HAL::millis();

by moving HAL_MINIMIZE_FEATURES's endif up and always continuing if less than 100ms has passed since the last time we ran bendyrule or dijkstra's.  I think the new method is better because it runs the code every 100ms instead of 100ms + processing time.

rmackay9 replied:
Ah, I see the issue is that you want to avoid a busy loop.  What we could do is instead calculate a delay time which is 1ms, 20ms or the time remaining until we need to run bendyrule or dijkstra's (i.e. 100ms - (now - avoidance_latest_ms))

magicrub replied:
good ideas for another PR

rmackay9 commented:
I've tested this in SITL and confirmed that it's working.

magicrub commented:
Summary of a test run:
Hardware: CubeBlack
OA_DB_SIZE=100
OA_DB_QUEUE=500
Scheduler rate = 50Hz
Camera - RealSense D435i
Mavlink packet OBSTACLE_DISTANCE (range 72 values) was coming into Proximity library at 15hz
This test includes my output branch where I was pushing the database out the mavlink port to view, so it's actually slightly higher than a normal user would see due to the extra outbound traffic

My camera is forward-looking only and is not working well outside. When in AUTO it tends to see random targets and spins around to look for other ways around, but can sometimes keep spinning as it continues to see noise. @rmackay9 will be testing on an rplidar soon. However, poor vs good camera/sensor doesn't matter for a cpu load test.

See attached log [OA_DB.zip](https://github.com/ArduPilot/ardupilot/files/3383219/OA_DB.zip)

LEGEND:
cpu = AP::scheduler().load_average() * 100.0 (so, it's a percent)
main = count of avoidance_thread() looped per second on average over 10 seconds
alg =  count of the BendyRuler algorithm looped per second on average over 10 seconds. This is triggered to run at 10Hz in the main avoidance_thread() loop
max = longest duration in ms of the whole main loop over 10 seconds
avg = average duration in ms of the whole main loop over 10 seconds

```
SITL (values with an almost unlimited cpu available)
APM: OA:cpu=0,main=56,alg=8,max=60.666,avg=26.203
APM: OA:cpu=0,main=56,alg=9,max=61.666,avg=22.211
APM: OA:cpu=0,main=57,alg=9,max=63.667,avg=23.215
APM: OA:cpu=0,main=56,alg=8,max=58.666,avg=22.425
APM: OA:cpu=0,main=56,alg=9,max=79.214,avg=21.893
APM: OA:cpu=0,main=57,alg=8,max=115.667,avg=22.797
APM: OA:cpu=0,main=56,alg=9,max=50.000,avg=25.464



LOG 00000002.bin
MSG, 68483530, OA:cpu=27,main=35,alg=7,max=558.803,avg=23.835
MSG, 78534334, OA:cpu=25,main=31,alg=8,max=156.371,avg=127.714
MSG, 88614536, OA:cpu=28,main=9,alg=7,max=213.918,avg=94.043
MSG, 98647515, OA:cpu=26,main=31,alg=9,max=129.808,avg=98.030
MSG, 108684486, OA:cpu=27,main=20,alg=9,max=127.940,avg=68.759
MSG, 128829194, OA:cpu=27,main=24,alg=9,max=126.448,avg=111.214
MSG, 138936417, OA:cpu=28,main=35,alg=9,max=122.610,avg=36.630
MSG, 149047714, OA:cpu=27,main=50,alg=9,max=91.045,avg=55.683
MSG, 159148599, OA:cpu=26,main=45,alg=9,max=118.854,avg=51.555
MSG, 169177566, OA:cpu=27,main=30,alg=9,max=131.308,avg=96.510
MSG, 179257719, OA:cpu=29,main=30,alg=9,max=117.023,avg=59.568
MSG, 189312542, OA:cpu=26,main=36,alg=9,max=134.048,avg=44.325
MSG, 199391085, OA:cpu=27,main=45,alg=9,max=143.798,avg=89.562
MSG, 209497759, OA:cpu=27,main=46,alg=9,max=161.931,avg=34.982
MSG, 219590443, OA:cpu=25,main=44,alg=9,max=124.851,avg=30.323

MSG, 64202770, PERF: 2/500 max=134940 min=19893 F=19975 sd=192076
MSG, 74202714, PERF: 0/500 max=20133 min=19870 F=20014 sd=16
MSG, 84206535, PERF: 0/500 max=20134 min=19865 F=19977 sd=16
MSG, 94202459, PERF: 0/500 max=20182 min=19818 F=20015 sd=17
MSG, 104205591, PERF: 0/500 max=20204 min=19797 F=19975 sd=17
MSG, 114202581, PERF: 0/500 max=20109 min=19891 F=19976 sd=15
MSG, 124206322, PERF: 0/500 max=20137 min=19863 F=20019 sd=15
MSG, 134205739, PERF: 0/500 max=20138 min=19867 F=19983 sd=20
MSG, 144206460, PERF: 0/500 max=20117 min=19880 F=19978 sd=13
MSG, 154202313, PERF: 0/500 max=20097 min=19903 F=20027 sd=15
MSG, 164202740, PERF: 0/500 max=20114 min=19892 F=20027 sd=17
MSG, 174202572, PERF: 0/500 max=20086 min=19914 F=20013 sd=10
MSG, 184207128, PERF: 0/500 max=20179 min=19822 F=19984 sd=17
MSG, 194205687, PERF: 0/500 max=20128 min=19882 F=19986 sd=17
MSG, 204205755, PERF: 0/500 max=20111 min=19888 F=19979 sd=18
MSG, 214202283, PERF: 0/500 max=20100 min=19900 F=20008 sd=16
MSG, 224202783, PERF: 0/500 max=20094 min=19906 F=20017 sd=15
```

I tested again at scheduler 400Hz and it didn't even run the thread hardly at all. Massive number of scheduler misses. Not worth continuing any tests on that. However, in AUTO mode it navigated the waypoints perfectly while in 400Hz but avoidance was basically not able to loop at all and thus waypoints executed as if avoidance it wasn't running

Summary: This is safe to use for 50Hz based vehicles with default OA_DB settings, namely OA_DB_100. I ran the Queue size larger to ensure we didn't drop any mavlink packets. I think it's worthwhile to bump that default up from 80 to 150 so we can at least get 2 full large packets in. There's no cpu difference, just more RAM and possibly a larger max loop if we're really getting a lot of data. I think that's worth it but we can evaluate that at a later time and keep it at 80 for now. We could also lower OA_DB_EXPIRE from 10 to 5, but again let's evaluate that later

WickedShell commented:
@magicrub out of curiosity was that a target of fmuv3 vs CubeBlack? Mostly relevant as CubeBlack is O3, while fmuv3 isn't...

magicrub commented:
@WickedShell --board CubeBlack

WickedShell commented:
> @WickedShell --board CubeBlack

Darn, was hoping for an easy win :)

lvale commented:
@magicrub how is the 435i providing data to the OA ?

magicrub commented:
@lvale It can be used via MAVROS but my tests are using a custom stand-alone python script Jaime and I wrote.

lvale commented:
> @lvale It can be used via MAVROS but my tests are using a custom stand-alone python script Jaime and I wrote.

@magicrub saw something similar on the intel docs, but you are connecting the Realsense camera to a tx1, tx2, rpi ?

magicrub commented:
on tx2

rmackay9 commented:
I tested this on my vehicle yesterday and I couldn't really see an improvement over what we have in master.  I think the issue is most likely not with this code but rather it comes from inaccuracies in other parts of the system including errors in the vehicle's position (GPS inaccuracies), heading (compass inaccuracy) and lidar inaccuracies (direction and lag).

I think we need two things:

- object position reporting to the GCS in a form that can be merged to master (i.e. the sending of the messages is prompted from the GCS_MAVLink object instead of from the Database itself.
- testing with a 1D lidar.  This may how a larger improvement with the database than the 360lidar shows.

In general, because of the memory and cpu usage of using the database I think we need to show that it works better than what we have in master before merging.

magicrub commented:
Well, I just drove around in AUTO with 2 waypoints with a database of 1000 and got zero schedule overruns and sd always <20. I don't have cpu load data from this run though

peterbarker commented:
On Fri, 12 Jul 2019, Randy Mackay wrote:

>  *  testing with a 1D lidar. This may how a larger improvement with the database than the 360lidar shows.

Having memory of the height of the terrain *in front of* the vehicle could 
be very, very useful.  We could, perhaps, get much better terrain 
following with it.

rmackay9 commented:
As discussed with @magicrub, I've got two commits in my repo here (https://github.com/rmackay9/rmackay9-ardupilot/commits/magicrub-oadb2) which send the adsb messages asynchronously.  It seems to work reasonably well in SITL with a database of up to about 200 points.

rmackay9 commented:
Here's a branch which is a squashed version of what @magicrub and I have done.

https://github.com/rmackay9/rmackay9-ardupilot/commits/magicrub-oadb3

rmackay9 commented:
@magicrub, I think I've fixed the two issues we discussed on the dev call and have updated my [magicrub-oadb3 branch](https://github.com/rmackay9/rmackay9-ardupilot/commits/magicrub-oadb3):

- ADSB messages are throttled to be no higher than the SRx_ADSB parameter value.  During testing on my copter I wasn't able to get it to send faster than about 37hz but my database size was only 100 and was testing in a confined area so it's possible the database simply wasn't filling up.
- Setting OA_DB_SIZE = 3000 fails more gracefully now.  At least the board boots up successfully although I haven't checked if we an arming failure is triggered (my guess is it doesn't trigger an arming failure).

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+    }
+
+    // if the filter value changed, constrain it to min/max and announce it
+    if (!is_equal(_database.filter_m, _database_filter_m_backup)) {
```
I think we've removed all debug..

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+            return;
+        }
+
+        free(_queue.items);
```
changed to use delete.

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        deinit_queue(false);
+        return;
+    } else if (_unable_to_allocate_memory) {
+        return;
```
fixed.

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+
+void AP_OADatabase::init()
+{
+    if (_initialised || _unable_to_allocate_memory || _database.size_param < (int32_t)_database.size_min) {
```
unable_to_allocate has been removed.

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        ::printf("DB::database_item_add() failed - FULL. Size: %d, filter=%.2fm\n",  _database.size,  _database.filter_m);
+        return;
+    }
+    _database.items[_database.count] = item;
```
this is out of date now so I think I'll mark it as resolved.

rmackay9 commented on libraries/AC_Avoidance/AP_OADatabase.cpp:
```diff
@@ -0,0 +1,467 @@
+        // compare item to all items in database. If found a similar item, update the existing, else add it as a new one
+        for (uint32_t i=0; i<_database.count && !found; i++) {
+            // skip last hit because we already checked it above
+            if (i != _database.last_hit && is_item_in_database(i, item)) {
```
last_hit handling has been removed..

magicrub commented:
@rmackay9 and I approve this PR. We'll leave it open another 2ish days for comment and testing before we merge

magicrub commented:
@jmachuca77 has tested this on some AION hardware and it works well. Ready for merge!

hoangthien94 commented:
> It can be used via MAVROS but my tests are using a custom stand-alone python script Jaime and I wrote.

Hi @magicrub, may I ask if the script is available somewhere? 

I am working on converting depth data from stereo images of the Realsense T265 to `OBSTACLE_DISTANCE` message and send to ArduPilot, but the documentation for this message seems very scarce. Any help would be greatly appreciated!

magicrub commented:
@hoangthien94 that packet seems straightforward to me. What about it needs explanation?

hoangthien94 commented:
There are two things, for the case of forward facing camera with a limited FOV:
- How to set symmetrical (left + center + right) array `distances` : Since the array  starts from the left in increment degrees to the right, I suppose `angle_offset`, `increment` (or `increment_f`) and the number of elements in `distances` will be of importance:
  - If I set any elements of `distances` to a value of `UINT16_MAX` for `unknown/not used`, would the next element (that contains valid value) be incremented by one, or two x `increment`? I imagine this might affect where the **straight ahead** (normally should be `distances[36]`) element lies.
  - Can the value of `angle_offset` be negative, or it must be positive [0,360)? 
- **Correct me if I am wrong on this one**: I stumbled upon this comment from https://github.com/PX4/avoidance/issues/460#issuecomment-519217565, which says that [`OBSTACLE_DISTANCE`](https://mavlink.io/en/messages/common.html#OBSTACLE_DISTANCE) is defined in the local frame, not the vehicle frame, even though it seems to contradict with the message definition. 
> Obstacle distances in front of the sensor,

magicrub commented:
@hoangthien94 answered on the other thread
